### PR TITLE
Support Apple Pay in cross-origin iframes with allow=payment attribute

### PR DESCRIPTION
**Pull request by Zeswen**
#### fead01e13ad2f5837324865b19f30607383d410c
<pre>
Support Apple Pay in cross-origin iframes with allow=payment attribute
<a href="https://bugs.webkit.org/show_bug.cgi?id=226345">https://bugs.webkit.org/show_bug.cgi?id=226345</a>

Reviewed by Dean Jackson.

This bug happens because there is a check for the parentDocument to have
the same security origin than the child document, contrary to what the
[Payment Request API spec from W3 states](<a href="https://www.w3.org/TR/payment-request/#using-with-cross-origin-iframes).">https://www.w3.org/TR/payment-request/#using-with-cross-origin-iframes).</a>
The solution is to remove the check for the same parentDocument origin.

* LayoutTests/http/tests/paymentrequest/payment-allow-attribute.https-expected.txt:
- Update tests according to new behavior
* Source/WebCore/Modules/applepay/PaymentSession.cpp:
(PaymentSession::canCreateSession):
- Remove check for same parent origin since we have allow payment check before

Canonical link: <a href="https://commits.webkit.org/262616@main">https://commits.webkit.org/262616@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/29ebb3c627747050737e9853ebb83816771effb2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2048 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2076 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2143 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/2973 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/2111 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/2022 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2156 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2118 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1871 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2066 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/1838 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1839 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/2823 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1824 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1815 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1728 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1884 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1863 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/2998 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1881 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/1682 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1808 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1826 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/511 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1988 "Built successfully") | | | 
| | | | |

### CODE REVIEW

**Review by marcoscaceres — COMMENTED** [review]
thanks @Zeswen! 

This lgtm from a spec conformance standpoint. However, as I'm not a Webkit reviewer I can't give it a proper approval. I've pinged folks to find out if Apple Pay can also handle this and will get another person to review it soon!

**Comment by ShreyTyagiZuora** [comment]
Thanks, @Zeswen for this fix. I have been trying to figure out why isn't `allow = "payment *"` not working and still throwing 
```SecurityError: Third-party iframes are not allowed to request payments unless explicitly allowed via Feature-Policy (payment)```

I did not find any Apple documentation that states third-party Iframes can't be used to render the ApplePay button, in hopes that it is indeed a bug. I see that some gateway has tried integrating it via the Iframe way and stated that it is not supported. https://docs.saferpay.com/home/integration-guide/iframe-integration-and-css

File: Source/WebCore/Modules/applepay/PaymentSession.cpp
Comment (grorg) [inline Source/WebCore/Modules/applepay/PaymentSession.cpp:1]: Note to other folks looking: you need to expand the diff upwards to see that we test for `isFeaturePolicyAllowedByDocumentAndAllOwners` with `Featurepolicy::Type::Payment`.

**Review by grorg — APPROVED** [review]
This looks good. @geoffreygaren or someone should confirm.

**Comment by Zeswen** [comment]
@geoffreygaren correct me if I am wrong but I believe the tests fails are unrelated and due to timeouts?

**Comment by marcoscaceres** [comment]
@Zeswen, want to give the is a quick rebase just to make the bots happy?

**Comment by marcoscaceres** [comment]
@Zeswen, want to give the is a quick rebase just to make the bots happy?

**Comment by Zeswen** [comment]
@marcoscaceres done!

**Comment by Zeswen** [comment]
Is there anything I can help with to accelerate the merging of this bugfix?

**Comment by marcoscaceres** [comment]
Thanks for pinging @Zeswen and sorry for the delay! Just triple checking it with a few folks it won't have any unintended side effects. Requested @geoffreygaren give it a final tick. 

@geoffreygaren, if it looks ok, could you "merge-queue" it?

**Comment by grorg** [comment]
I think it's ok to merge.

**Comment by grorg** [comment]
Why is this taking so long to merge?

**Comment by Zeswen** [comment]
@marcoscaceres rebased!

**Comment by marcoscaceres** [comment]
Yikes. It went all red. I wonder if it's getting upset because there are multiple commits? @Zeswen, you might need to squash them.

**Comment by Zeswen** [comment]
@marcoscaceres squashed now, sorry about that! I was trying to avoid downloading the whole repo but no luck 😅

**Comment by grorg** [comment]
Adding the merge label again. Fingers crossed!

**Review by grorg — APPROVED** [review]

**Comment by grorg** [comment]
As usual, I have no idea why our tools don't want to do things. I had already reviewed the patch. I added a new review in the hope that unblocks things.

**Comment by marcoscaceres** [comment]
@Zeswen, looks like the commit message no longer complies with the expected format. It should look like:
<pre>
Support Apple Pay in cross-origin iframes with allow=payment attribute
https://bugs.webkit.org/show_bug.cgi?id=226345

Reviewed by NOBODY (OOPS!).

This bug happens because there is a check for the parentDocument to have 
the same security origin than the child document, contrary to what the 
[Payment Request API spec from W3 states](https://www.w3.org/TR/payment-request/#using-with-cross-origin-iframes).
The solution is to remove the check for the same parentDocument origin.

* Source/WebCore/Modules/applepay/PaymentSession.cpp:
(PaymentSession::canCreateSession):
- Remove check for same parent origin since we have allow payment check before

</pre>

@Zeswen, have you been using `git webkit pr` to push the PR updates? If not, that might be what's causing the issues.

The usual workflow is:

```
git commit --amend files/changed/during/review.cpp
git webkit pr
```

IIRC, when you call `commit --amend`, it should rewrite the commit message properly for you in the expected format.

**Comment by Zeswen** [comment]
@marcoscaceres @grorg updated the PR description and commit with the proper messages!

Thank you for the insightful comment on how to fix this @marcoscaceres 🙂 
Unfortunately, since I created the fork without `git webkit pr`, the tool doesn't want to cooperate now 😅 Hopefully modifying the commit message manually does the trick!

**Comment by marcoscaceres** [comment]
Great! Glad that worked out and it got merged.

**Comment by Janving** [comment]
Great Job! Do you know when this will be released to the public users?

**Comment by daelmaak** [comment]
Thanks for the fix. Any plan as to when this fix makes it to production?

**Comment by bmeeder22** [comment]
This fix is currently in the Safari Technology Preview! https://developer.apple.com/safari/technology-preview/

**Comment by Mastrangelis** [comment]
Hi, do you know if this got included in the 16.5 release? I tried it on 16.5 and still getting the same error even though it works just fine in the Safari Technoloy Preview

**Comment by TheMechanic** [comment]
The last Safari preview integrates this fix (since preview 167 : Added support for Apple Pay in cross-origin iframes with allow=payment attribute ([262616@main](https://commits.webkit.org/262616@main))), but I'm not able to make it work :/
- Parent page and iframe are served with https
- iframe domain is verified with ApplePay
- I have a card registered on ApplePay (and it works on other websites)

I don't get what's the issue... any idea ?

thank you :)

**Comment by ryan-lambert-IEX** [comment]
@TheMechanic are you using stripe per chance? I couldn't get it to work in our Stripe implementation but I did get it to work using this other demo https://developer.jupiterhq.com/products/payments/credit-card/applepay-interactive-demo

I could be totally off base, but I created this issue with Stripe https://github.com/stripe/stripe-js/issues/484 so we'll see!

**Pull request merged**